base/loading.jl

```diff
@@ -525,7 +527,23 @@ function pkgdir(m::Module, paths::String...)
     rootmodule = moduleroot(m)
     path = pathof(rootmodule)
     path === nothing && return nothing
-    return joinpath(dirname(dirname(path)), paths...)
+    original = path
+    path = dirname(path)
+    if splitpath(path)[end] == "src"
+        path = dirname(path)
+    elseif "ext" in splitpath(path)
+        # extensions can reside at `../ext/FooExt.jl` or `../ext/FooExt/FooExt.jl`
+        if splitpath(path)[end] == "ext"
+            path = dirname(path)
+        else
+            path = dirname(path)
+            splitpath(path)[end] == "ext" || error("Unexpected path structure for extension module: $original")
+            path = dirname(path)
+        end
+    else
+        error("Unexpected path structure for module source: $original")
+    end
```

Comment:
Could use `splitdir` to simplify the logic a bit
```suggestion
    path, base = splitdir(path)
    if base == "src"
        # default
    elseif base == "ext"  # extension in `../ext/FooExt.jl`
        # source code
    elseif basename(path) == "ext"  # extension in `../ext/FooExt/FooExt.jl`
        path = dirname(path)
    else
        error("Unexpected path structure for module source: $original")
    end
```
Or even just this (less resilient but simpler)
```suggestion
    path, base = splitdir(path)
    if basename(path) == "ext"
        # extension in `../ext/FooExt/FooExt.jl` so we need to go up another level
        path = dirname(path)
    end
```

Reply:
Thanks. Pushed with minor tweaks.